src/mscorlib/shared/System/Memory.cs

```diff
@@ -139,7 +139,9 @@ public Memory<T> Slice(int start)
 
             if (_index < 0)
                 return new Memory<T>((OwnedMemory<T>)_arrayOrOwnedMemory, (_index & RemoveOwnedFlagBitMask) + start, _length - start);
-            return new Memory<T>((T[])_arrayOrOwnedMemory, _index + start, _length - start);
+            if (_arrayOrOwnedMemory != null)
+                return new Memory<T>((T[])_arrayOrOwnedMemory, _index + start, _length - start);
```

Comment:
Same question as here:
https://github.com/dotnet/corefx/pull/25009#pullrequestreview-73679606

(In the future, I'd suggest putting up just one of the PRs, incorporating any feedback, then putting up the other PR.  Otherwise feedback ends up either getting split or duplicated between them.)

Reply:
> How does the JIT do at eliminating the null check in the inlined Memory ctor?  For that matter, how does it do at eliminating the range checks that also shouldn't be necessary?

It looks like the JIT eliminates the extra null check (testing on a sample app and confirming with Slice before/after). It doesn't eliminate the range checks though.

![image](https://user-images.githubusercontent.com/6527137/32476847-fb744fce-c32e-11e7-8715-176065318e48.png)

> I'm wondering if we should have a private ctor that let's us bypass those additional checks, especially since this is aggressively inlined.

How would we write such a ctor overload? We can't change the ctor signature by renaming it or changing the arguments.

Reply:
> We can't change the ctor signature by renaming it or changing the arguments.

A dummy bool argument?

Reply:
You can have ctor that just takes `object` - it will save the cast as well.

Reply:
Even better.

Reply:
Will use the private ctor after this is merged:
https://github.com/dotnet/coreclr/pull/14906